src/csharp/Microsoft.Spark/Sql/StorageLevel.cs

```diff
@@ -0,0 +1,80 @@
+    /// data in memory in a JAVA-specific serialized format, and whether to replicate the RDD
+    /// partitions on multiple nodes. Also contains static properties for some commonly used
+    /// storage levels, MEMORY_ONLY.
+    /// </summary>
+    public sealed class StorageLevel : IJvmObjectReferenceProvider
+    {
+        private readonly JvmObjectReference _jvmObject;
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        internal StorageLevel(JvmObjectReference jvmObject)
+        {
+            UseDisk = (bool)jvmObject.Invoke("useDisk");
+            UseMemory = (bool)jvmObject.Invoke("useMemory");
+            UseOffHeap = (bool)jvmObject.Invoke("useOffHeap");
+            Deserialized = (bool)jvmObject.Invoke("deserialized");
+            Replication = (int)jvmObject.Invoke("replication");
+            _jvmObject = jvmObject;
+        }
+
+        public StorageLevel(
+            bool useDisk,
+            bool useMemory,
+            bool useOffHeap,
+            bool deserialized,
+            int replication = 1)
+        {
+            UseDisk = useDisk;
+            UseMemory = useMemory;
+            UseOffHeap = useOffHeap;
+            Deserialized = deserialized;
+            Replication = replication;
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(
+                "org.apache.spark.storage.StorageLevel",
+                UseDisk,
+                UseMemory,
+                UseOffHeap,
+                Deserialized,
+                Replication);
+        }
+
+        public bool UseDisk { get; private set; }
+        public bool UseMemory { get; private set; }
+        public bool UseOffHeap { get; private set; }
+        public bool Deserialized { get; private set; }
+        public int Replication { get; private set; }
+        public static Lazy<StorageLevel> DISK_ONLY { get; } =
+            new Lazy<StorageLevel>(() => new StorageLevel(true, false, false, false));
+        public static Lazy<StorageLevel> DISK_ONLY_2 { get; } =
+            new Lazy<StorageLevel>(() => new StorageLevel(true, false, false, false, 2));
+        public static Lazy<StorageLevel> MEMORY_ONLY { get; } =
+            new Lazy<StorageLevel>(() => new StorageLevel(false, true, false, false));
+        public static Lazy<StorageLevel> MEMORY_ONLY_2 { get; } =
+            new Lazy<StorageLevel>(() => new StorageLevel(false, true, false, false, 2));
+        public static Lazy<StorageLevel> MEMORY_AND_DISK { get; } =
+            new Lazy<StorageLevel>(() => new StorageLevel(true, true, false, false));
+        public static Lazy<StorageLevel> MEMORY_AND_DISK_2 { get; } =
+            new Lazy<StorageLevel>(() => new StorageLevel(true, true, false, false, 2));
+        public static Lazy<StorageLevel> OFF_HEAP { get; } =
+            new Lazy<StorageLevel>(() => new StorageLevel(true, true, true, false, 1));
```

Comment:
these should still be `public static StorageLevel`.  You'll need to do something like `public static Lazy<StorageLevel> DISK_ONLY => _disk_only.Value`

On a related note.  Should we hard code these true/false/# values here or should we call the `scala` object instead ?

Reply:
@suhsteve Please check if this makes sense, the option 2 is calling the scala object directly but `CallConstructor` method gives an error there.

Reply:
For option 2 you'll need use `CallStatic` instead of `CallConstructor`.  Something like...

```csharp
private static readonly string s_storageLevelClassName = "org.apache.spark.storage.StorageLevel"

...

private static CallbackServer s_diskOnly;
internal static CallbackServer DISK_ONLY
{
    get => s_diskOnly ??= new StorageLevel(
        (JvmObjectReference)SparkEnvironment.JvmBridge.CallStaticJavaMethod(
            s_storageLevelClassName,
            "DISK_ONLY"));
}
```